iris-mpc-cpu/src/execution/hawk_main.rs

```diff
@@ -482,37 +482,98 @@ impl HawkActor {
     }
 
-    async fn compute_buckets(&self, session: &mut Session, side: usize) -> Result<VecBuckets> {
+    async fn compute_buckets(
+        &self,
+        sessions: &BothEyes<Vec<HawkSessionRef>>,
+        side: usize,
+    ) -> Result<VecBuckets> {
         let translated_thresholds = Self::calculate_threshold_a(self.args.n_buckets);
-        let bucket_result_shares = compare_threshold_buckets(
-            session,
-            translated_thresholds.as_slice(),
-            self.distances_cache[side].as_slice(),
-        )
-        .await?;
+        let n_sessions = sessions[side].len();
+        let distances = self.distances_cache[side].as_slice();
+        let chunk_size = distances.len() / n_sessions;
+        let remainder = distances.len() % n_sessions;
+
+        let mut start = 0;
+        let mut slices_data = Vec::with_capacity(n_sessions);
+        for i in 0..n_sessions {
+            let size = chunk_size + (i < remainder) as usize;
+            let end = start + size;
+            slices_data.push(distances[start..end].to_vec());
+            start = end;
+        }
+
+        let (tx, mut rx) = mpsc::unbounded_channel::<(usize, Vec<Share<u32>>)>();
+        let tasks = sessions[side]
+            .iter()
+            .cloned()
+            .zip(slices_data.into_iter())
+            .enumerate()
+            .map(|(i_session, (session_ref, slice))| {
+                let thresholds = translated_thresholds.clone();
+                let tx = tx.clone();
+                async move {
+                    let mut session = session_ref.write().await;
+                    let shares = compare_threshold_buckets(
+                        &mut session.aby3_store.session,
+                        thresholds.as_slice(),
+                        &slice,
+                    )
+                    .await?;
+                    tx.send((i_session, shares))?;
+                    Ok(())
+                }
+            });
+
+        parallelize(tasks).await?;
+        drop(tx);
+
+        let mut shares_acc = vec![Vec::new(); n_sessions];
+        while let Some((i_session, shares)) = rx.recv().await {
+            shares_acc[i_session] = shares;
+        }
+
+        let mut bucket_result_shares = Vec::new();
+        for shares in shares_acc {
+            bucket_result_shares.extend(shares);
+        }
 
-        let buckets = open_ring(session, &bucket_result_shares).await?;
+        let mut session0 = sessions[side][0].write().await;
+        // Open all per-session bucket shares and aggregate into final per-threshold counts
+        let opened = open_ring(&mut session0.aby3_store.session, &bucket_result_shares).await?;
+        let n_buckets = translated_thresholds.len();
+        let mut buckets = vec![0u32; n_buckets];
+        for chunk in opened.chunks(n_buckets) {
+            for (i, &count) in chunk.iter().enumerate() {
+                buckets[i] = buckets[i].saturating_add(count);
+            }
+        }
```

Comment:
I guess this leaks a bit more than strictly needed, as we could add up the shares before opening them.

Reply:
something like (untested)
```

        let n_buckets = translated_thresholds.len();
        let mut share_buckets = vec![Share::<u32>::default(); n_buckets];
        while let Some((i_session, shares)) = rx.recv().await {
            for (share, bucket) in izip!(shares, share_buckets.iter_mut()) {
                *bucket += share;
            }
        }

        let mut session0 = sessions[side][0].write().await;
        // Open aggregated buckets
        let opened = open_ring(&mut session0.aby3_store.session, &share_buckets).await?;
        Ok(opened)
```

Reply:
Right. So instead of 

```rust
        let mut bucket_result_shares = Vec::new();
        for shares in shares_acc {
            bucket_result_shares.extend(shares);
        }
```

it could do something like (draft):

```rust
        let mut bucket_result_shares = vec![EMPTY_SHARE; n_bucket];
        for shares in shares_acc {
            for i in 0..n_buckets {
                bucket_result_shares[i] += shares[i];
            }
        }
```

This is similar to this line:

https://github.com/worldcoin/iris-mpc/blob/92083051af3586320f3764082db9b61491abcde2/iris-mpc-cpu/src/protocol/ops.rs#L139

Reply:
Went with @naure 's suggestion here